# Analyzing Online Ticket Sales with Amazon Redshift

In this project, I will be accessing data stored in Amazon Redshift, a data warehouse product that is part of Amazon Web Services. More specifically, I'll be analyzing sales activity from a fictional ticketing website where users both buy and sell tickets online for sporting events, shows, and concerts ([source](https://docs.aws.amazon.com/redshift/latest/dg/c_sampledb.html)).

## Aim of The Project
Which users should we target for our advertising purposes?
- There are several users, who have advertised or have put tickets up for sale but haven't been able to sell them effectively. 
- By recognizing these users, we can specifically target them through advertizing. 

## Tech Stack
- I'll be using SQL to query the Redshift database as well as to manipulate the data.
- I'll also use Python to visualize this data na dform interesting conclusions.

## Data Exploration Using SQL
Let's start by exploring the events data.

In [ ]:
-- List all the events
SELECT * FROM event

eventid  venueid  catid  dateid              eventname  \
0         423      303      8    1827  La Damnation de Faust   
1        2099      211      7    1827            The Bacchae   
2        6640       38      9    1827         Herbie Hancock   
3        8621        9      9    1827               Hot Chip   
4        3979      127      9    1828          Alkaline Trio   
...       ...      ...    ...     ...                    ...   
8793     6034       45      9    2191                    War   
8794     6783       60      9    2191             The Police   
8795     6857       18      9    2191           Judas Priest   
8796     7192       54      9    2191     Lindsey Buckingham   
8797     8168       70      9    2191      Little River Band   

                     starttime  
0    2008-01-01 19:30:00+00:00  
1    2008-01-01 19:30:00+00:00  
2    2008-01-01 19:30:00+00:00  
3    2008-01-01 19:00:00+00:00  
4    2008-01-02 14:30:00+00:00  
...                        ...  
8793 2008-12-31 14:00:00+00:00  
8794 2008-12-31 15:00:00+00:00  
8795 2008-12-31 14:00:00+00:00  
8796 2008-12-31 19:30:00+00:00  
8797 2008-12-31 15:00:00+00:00  

[8798 rows x 6 columns]

- There are around 9000 different events for year 2008 in this database. 
- This is linking up to several other tables in the warehouse, such as venue, category and date. 
- Using SQL's 'INNER JOIN' capability, let's join things up to see a much detailed table. 

In [ ]:
-- joining venue and category tables to the events table
SELECT *
FROM event
INNER JOIN venue USING(venueid)
INNER JOIN category USING(catid)
INNER JOIN date USING(dateid)
LIMIT 100;

dateid  catid  venueid  eventid           eventname  \
0     1827      6      238     1217          Mamma Mia!   
1     1827      7      207     2811    Spring Awakening   
2     1827      9       72     6711   Smashing Pumpkins   
3     1827      9        6     7747           K.D. Lang   
4     1828      7      209     2920             Macbeth   
..     ...    ...      ...      ...                 ...   
95    1840      7      214     1970       The Caretaker   
96    1840      7      209     2907  Look Back in Anger   
97    1840      7      217     3339            Thurgood   
98    1840      9       28     4476         Temptations   
99    1840      9       48     5966    Rick Springfield   

                   starttime                  venuename        venuecity  \
0  2008-01-01 20:00:00+00:00      Winter Garden Theatre    New York City   
1  2008-01-01 15:00:00+00:00  Bernard B. Jacobs Theatre    New York City   
2  2008-01-01 19:00:00+00:00   Cleveland Browns Stadium        Cleveland   
3  2008-01-01 15:00:00+00:00    New York Giants Stadium  East Rutherford   
4  2008-01-02 19:30:00+00:00              Booth Theatre    New York City   
..                       ...                        ...              ...   
95 2008-01-14 15:00:00+00:00    George Gershwin Theatre    New York City   
96 2008-01-14 19:00:00+00:00              Booth Theatre    New York City   
97 2008-01-14 19:30:00+00:00             Hilton Theatre    New York City   
98 2008-01-14 19:00:00+00:00    American Airlines Arena            Miami   
99 2008-01-14 19:00:00+00:00                 HSBC Arena          Buffalo   

   venuestate  venueseats  catgroup   catname  \
0          NY         0.0     Shows  Musicals   
1          NY         0.0     Shows     Plays   
2          OH     73200.0  Concerts       Pop   
3          NJ     80242.0  Concerts       Pop   
4          NY         0.0     Shows     Plays   
..        ...         ...       ...       ...   
95         NY         0.0     Shows     Plays   
96         NY         0.0     Shows     Plays   
97         NY         0.0     Shows     Plays   
98         FL         0.0  Concerts       Pop   
99         NY         0.0  Concerts       Pop   

                            catdesc                   caldate  day  week  \
0                   Musical theatre 2008-01-01 00:00:00+00:00  WE      1   
1           All non-musical theatre 2008-01-01 00:00:00+00:00  WE      1   
2   All rock and pop music concerts 2008-01-01 00:00:00+00:00  WE      1   
3   All rock and pop music concerts 2008-01-01 00:00:00+00:00  WE      1   
4           All non-musical theatre 2008-01-02 00:00:00+00:00  TH      1   
..                              ...                       ...  ...   ...   
95          All non-musical theatre 2008-01-14 00:00:00+00:00  TU      3   
96          All non-musical theatre 2008-01-14 00:00:00+00:00  TU      3   
97          All non-musical theatre 2008-01-14 00:00:00+00:00  TU      3   
98  All rock and pop music concerts 2008-01-14 00:00:00+00:00  TU      3   
99  All rock and pop music concerts 2008-01-14 00:00:00+00:00  TU      3   

    month  qtr  year  holiday  
0   JAN      1  2008     True  
1   JAN      1  2008     True  
2   JAN      1  2008     True  
3   JAN      1  2008     True  
4   JAN      1  2008    False  
..    ...  ...   ...      ...  
95  JAN      1  2008    False  
96  JAN      1  2008    False  
97  JAN      1  2008    False  
98  JAN      1  2008    False  
99  JAN      1  2008    False  

[100 rows x 20 columns]

## Analyzing Duplicates
-- There's a `starttime` column coming from the `event` table and there's also a `caldate` column, coming from the `date` table. Let's see what's up with this.

In [ ]:
SELECT
	CASE WHEN DATE(caldate) = DATE(starttime) THEN True ELSE False END AS same_date,
	COUNT(*)
FROM event
INNER JOIN date USING(dateid)
GROUP BY 1

same_date  count
0      False    703
1       True   8095

- So, there are 703 recirds, where the 'caldate' column does not match up with the 'starttime' column.
- Now, let's find out the maximum difference (in hours) between the 'caldate' time and 'startime' time.

In [ ]:
-- Using the DATEDIFF function here to calculate the diff in hours
SELECT MAX(DATEDIFF('hour', caldate, starttime))
FROM event
INNER JOIN date USING(dateid)

max
0   20

- The Maximum difference is just 20 hours. 
- However, for our analysis, we can just leave this as is because we're more interested in the 'users' data. 

## Adding in More Variables
- Let's see how much events are happening in different cities.

In [ ]:
SELECT 
	venuecity,
    COUNT(*) AS num_events
FROM event
INNER JOIN venue USING(venueid)
GROUP BY 1
ORDER BY 2 DESC

venuecity  num_events
0   New York City        2647
1     Los Angeles         312
2       Las Vegas         300
3         Chicago         209
4   San Francisco         194
..            ...         ...
74       Montreal          27
75         Newark          27
76         Irving          25
77        Sunrise          24
78        Calgary          18

[79 rows x 2 columns]

- New York City hosted the most number of events. 
- Calgary, on the other hand, hosted the least amount of events. 

## Data Viz
- Let's use Python's plotly library for visulaizing this data. 

In [ ]:
# Import plotly
import plotly.express as px

# Bar plot
px.bar(events_per_city, x = 'venuecity', y = 'num_events')

## Explore listings and sales

- There are two tables that we haven't touched so far (sales & listings).
- These are important if we want to calucate revenue (money made from sales)
- Let's explore them here

In [ ]:
-- show 100 listing records
SELECT * FROM listing LIMIT 100;

listid  sellerid  eventid  dateid  numtickets  priceperticket  totalprice  \
0     2092     42560     8609    1827          22           194.0      4268.0   
1     9242     31448     7441    1827          14            88.0      1232.0   
2     9730      8250     2772    1827          20           146.0      2920.0   
3    11176     39919     5362    1827          12           107.0      1284.0   
4    11615     15795     5157    1827           4           221.0       884.0   
..     ...       ...      ...     ...         ...             ...         ...   
95   10309     13587     3797    1828           4            96.0       384.0   
96   10900     33297     1281    1828          12           217.0      2604.0   
97   10970     31450     7546    1828           1           105.0       105.0   
98   21681     47532     5346    1828          28            67.0      1876.0   
99   22412     42508     3139    1828           4           185.0       740.0   

                    listtime  
0  2008-01-01 05:49:06+00:00  
1  2008-01-01 02:39:54+00:00  
2  2008-01-01 09:46:40+00:00  
3  2008-01-01 10:18:43+00:00  
4  2008-01-01 06:02:18+00:00  
..                       ...  
95 2008-01-02 01:51:41+00:00  
96 2008-01-02 05:12:22+00:00  
97 2008-01-02 06:59:22+00:00  
98 2008-01-02 07:59:20+00:00  
99 2008-01-02 06:09:05+00:00  

[100 rows x 8 columns]

In [ ]:
-- show 100 sales records
SELECT * FROM sales LIMIT 100;

salesid  listid  sellerid  buyerid  eventid  dateid  qtysold  pricepaid  \
0     65082   73790     20429      451     1150    1827        4      472.0   
1     84644   96603      6051     1312     6641    1828        2      810.0   
2    160330  212426     37430      447     6884    1828        4     8672.0   
3     16536   17885     33096     1739     7443    1829        2      342.0   
4     64837   73489     21131     2375     4488    1829        2      156.0   
..      ...     ...       ...      ...      ...     ...      ...        ...   
95   111878  128148     10913     1462     2524    1835        1      335.0   
96   112994  129445      6289      261     2472    1835        2      472.0   
97   115182  131852     18717     5993     3884    1835        1      351.0   
98   116332  133228     36608     6824     2676    1835        1      204.0   
99   142399  164770     40822     2245     8069    1835        2       64.0   

    commission                  saletime  
0        70.80 2008-01-01 06:06:57+00:00  
1       121.50 2008-01-02 09:31:15+00:00  
2      1300.80 2008-01-02 03:20:32+00:00  
3        51.30 2008-01-03 06:36:27+00:00  
4        23.40 2008-01-03 05:58:02+00:00  
..         ...                       ...  
95       50.25 2008-01-09 10:20:55+00:00  
96       70.80 2008-01-09 01:51:41+00:00  
97       52.65 2008-01-09 06:27:11+00:00  
98       30.60 2008-01-09 05:10:56+00:00  
99        9.60 2008-01-09 08:42:42+00:00  

[100 rows x 10 columns]

- Let's see if multiple sales can happen for the same listing.
- In other words, are there multiple sales associated with the same listid?
- Here, I'll be using SQL's 'HAVING' clause.

In [ ]:
    SELECT 
        listid,
        COUNT(*) AS number_of_sales
    FROM listing
	INNER JOIN sales USING(listid)
    GROUP BY 1
    HAVING COUNT(*) > 1;

listid  number_of_sales
0       18088                2
1       18250                3
2       45435                3
3       56766                3
4       57663                2
...       ...              ...
53050  172518                2
53051   60494                2
53052   45075                2
53053  101779                2
53054   92363                2

[53055 rows x 2 columns]

- Using the abouve query as a Common Table Expression, let's count how many sales happened once as well as 'more than once'.

In [ ]:
WITH listings_with_sales AS (
    SELECT 
        listid,
        COUNT(*) AS number_of_sales
    FROM listing
	INNER JOIN sales USING(listid)
    GROUP BY 1
)
SELECT
	number_of_sales,
    COUNT(*)
FROM listings_with_sales
GROUP BY 1
ORDER BY 1

number_of_sales  count
0                1  48029
1                2  36570
2                3  14665
3                4   1808
4                5     12

- It turns out the bulk of listings have only one sale associated with them. 
- However, There are 12 listings that had 5 sales.

## User Data
- It's time to put some names to numbers. 
- Let's see how many tickets (qty) did each user sell.

In [ ]:
SELECT
	sellerid,
    username,
    (firstname ||' '|| lastname) as name, -- this is a shortcut way to use the CONCAT method
	city,
    sum(qtysold)
FROM sales
INNER JOIN date USING(dateid)
INNER JOIN users ON sales.sellerid = users.userid
WHERE year = 2008
GROUP BY 1, 2, 3, 4
ORDER BY 5 desc
LIMIT 5

sellerid  username              name         city  sum
0     48950  TUT90BHI        Nayda Hood       Frisco   46
1     20029  RPM45HGY      Drew Mcguire    Lancaster   41
2     19123  DZW00VOQ     Scott Simmons       Carson   41
3     36791  DCE77DOA  Emerson Delacruz  Springfield   40
4     41579  QXQ28WLA   Harrison Durham     Anaconda   39

- Similarly, we can find the most active buyer on the site in 2008.
- These are the buyers, who bought the tickets from the sellers. 

In [ ]:
select
	buyerid,
    username,
    (firstname ||' '|| lastname) as name,
	city,
    sum(qtysold)
FROM sales
INNER JOIN date USING(dateid)
INNER JOIN users ON sales.buyerid = users.userid
WHERE year = 2008
GROUP BY 1, 2, 3, 4
ORDER BY 5 desc
LIMIT 5

buyerid  username           name          city  sum
0     8933  CNF70VPH  Jerry Nichols    Middlebury   67
1     1298  EDB46JXK  Kameko Bowman   Newburyport   64
2     3797  KTV94TWB  Armando Lopez        Pomona   64
3     5002  CBC51API  Kellie Savage  Falls Church   63
4     4064  IQS59DPH   Kadeem Blair       Anaheim   60

## Analyzing the Prices
- Let's see if there's a big difference in average sales price for different categories of events. We're looking at actual sales here, not listings!
- Here, we're using some summary statistics to get agood grasp pn the data.

In [ ]:
SELECT
	catgroup,
    AVG(pricepaid/qtysold) AS avg_ticket_price,
    MEDIAN(pricepaid/qtysold) AS median_ticket_price
FROM sales
INNER JOIN event USING(eventid)
INNER JOIN category USING(catid)
GROUP BY 1

catgroup  avg_ticket_price  median_ticket_price
0  Concerts        333.755006                229.0
1     Shows        336.982704                232.0

Some observations: 
- Median is lower than the average. This means that there are outliers (high sale prices).
- The graph is rightly Skewed. 

## Time Discrepencies
Are there listings where the sale happened before the listing?

In [ ]:
SELECT COUNT(*)
FROM listing
INNER JOIN sales USING (listid)
WHERE listtime > saletime

count
0   2965

- Turns out there are instances where this happened. 
- This must be bad data! Let's keep these out when figuring out the shortest and longest time to get a listing sold.
- I'll be usind the DATEDIFF function here. 

In [ ]:
WITH tts AS (
    SELECT DATEDIFF('seconds', listtime, saletime) AS time_to_sell
    FROM sales
    INNER JOIN listing USING(listid)
    WHERE listtime < saletime
)
SELECT 
	MIN(time_to_Sell) AS shortest_time_to_sell_in_seconds,
    MAX(time_to_Sell)/3600/24 AS longest_time_to_sell_in_days
FROM tts

shortest_time_to_sell_in_seconds  longest_time_to_sell_in_days
0                                17                            60

## Finding users that should advertise

Suppose we, as owners of the ticketing website, want to target certain users with the suggestion to advertise their listings, so they have a higher chance of actually getting sales. Let's build up a list of users that had the most outstanding listings in terms of price per ticket.

- First, let's see which users did not sell any tickets. 
- I'll be using the LEFT join
- Also, using the COALESCE funtion I'll replace NUll values with 0. 

In [ ]:
SELECT  listid,
 	    COALESCE(qtysold, 0)
    FROM listing
		LEFT JOIN sales USING(listid)
WHERE qtysold IS NULL

listid  coalesce
0        8319         0
1       60816         0
2      122714         0
3      160788         0
4      220953         0
...       ...       ...
91408   97982         0
91409   37567         0
91410  114963         0
91411  231458         0
91412    1486         0

[91413 rows x 2 columns]

- Now, let's see which users havenot sold tickets yet. 
- We will calculate their 'unrelaized earnings', which is a fancy way of saying potential revenue.
- In our advertizing campaign, we can target these users to help them sell. 

In [ ]:
WITH listings_with_sales AS (
    SELECT 
        listid,
 	    listing.sellerid,
        numtickets AS tickets_listed,
        priceperticket,
		SUM(COALESCE(qtysold, 0)) AS tickets_sold
    FROM listing
	LEFT JOIN sales USING(listid)
    GROUP BY 1, 2, 3, 4
)
SELECT 
	sellerid,
    (firstname ||' '|| lastname) as name,
	SUM((tickets_listed - tickets_sold) * priceperticket) AS unrealized_sales
FROM listings_with_sales lws
INNER JOIN users ON lws.sellerid = users.userid
GROUP BY 1, 2
ORDER BY 3 DESC
LIMIT 100

sellerid              name  unrealized_sales
0      25428      Jaime Wagner           58395.0
1      24896       Macey Ortiz           53086.0
2      49322    Dustin Vincent           50914.0
3      36926     Audrey Barber           50345.0
4      45819     Kelly Barrett           49826.0
..       ...               ...               ...
95     35926    Ulysses Kinney           38895.0
96     45372  Lysandra Sanchez           38862.0
97     48188    Caesar Parrish           38847.0
98     25373       Jakeem Byrd           38847.0
99      1759     Lane Franklin           38740.0

[100 rows x 3 columns]

- Looks like Jaime Wagner had 58k of unrealized sales!
- Let's limit this list to people who have more than 40K in unrealized sales. 

In [ ]:
WITH listings_with_sales AS (
    SELECT 
        listid,
 	    listing.sellerid,
        numtickets AS tickets_listed,
        priceperticket,
		SUM(COALESCE(qtysold, 0)) AS tickets_sold
    FROM listing
	LEFT JOIN sales USING(listid)
    GROUP BY 1, 2, 3, 4
)
SELECT 
	sellerid,
    (firstname ||' '|| lastname) as name,
	SUM((tickets_listed - tickets_sold) * priceperticket) AS unrealized_sales
FROM listings_with_sales lws
INNER JOIN users ON lws.sellerid = users.userid
GROUP BY 1, 2
HAVING unrealized_sales > 40000
ORDER BY 3 DESC
LIMIT 100

sellerid             name  unrealized_sales
0      25428     Jaime Wagner           58395.0
1      24896      Macey Ortiz           53086.0
2      49322   Dustin Vincent           50914.0
3      36926    Audrey Barber           50345.0
4      45819    Kelly Barrett           49826.0
..       ...              ...               ...
68     20845    Lance Camacho           40152.0
69     12380        Amy Cooke           40129.0
70     47247     Hayley Lloyd           40122.0
71     35355  Kathleen Bonner           40122.0
72     39977       Ocean Rios           40057.0

[73 rows x 3 columns]

# Conclusion
- Here is the list of the people that we can target for advertizing purposes.
- Let's convert this list to a bar graph. 

In [ ]:
# Bar plot
px.bar(df_users, x = 'name', y = 'unrealized_sales')

## Extra Bit of Detail
- Let's visualize sales over time and separate them based on category.
- This can help us to determine the seasonality of the sales.
- Using the DATE_TRUNC function to extract the 'week' from sales date. 

In [ ]:
-- Show total pricepaid on a weekly basis
SELECT
	catgroup,
    DATE_TRUNC('week', saletime) AS sales_week,
    SUM(pricepaid) AS total_sales
FROM sales
INNER JOIN event USING(eventid)
INNER JOIN category USING(catid)
GROUP BY 1, 2
ORDER BY 2;

catgroup                sales_week  total_sales
0    Concerts 2007-12-31 00:00:00+00:00     146716.0
1       Shows 2007-12-31 00:00:00+00:00      98275.0
2    Concerts 2008-01-07 00:00:00+00:00     420710.0
3       Shows 2008-01-07 00:00:00+00:00     313664.0
4       Shows 2008-01-14 00:00:00+00:00     532037.0
..        ...                       ...          ...
101     Shows 2008-12-15 00:00:00+00:00     419135.0
102     Shows 2008-12-22 00:00:00+00:00     245468.0
103  Concerts 2008-12-22 00:00:00+00:00     289723.0
104     Shows 2008-12-29 00:00:00+00:00      37838.0
105  Concerts 2008-12-29 00:00:00+00:00      44918.0

[106 rows x 3 columns]

In [ ]:
#Plotting a line graph
import plotly.express as px
px.line(sales_over_week, x = 'sales_week', y = 'total_sales', color = 'catgroup')